moveit_core/planning_scene/test/test_multi_threaded.cpp

```diff
@@ -0,0 +1,135 @@
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Jens Petit */
+
+#include <moveit/robot_model/robot_model.h>
+#include <moveit/robot_state/robot_state.h>
+#include <moveit/utils/robot_model_test_utils.h>
+#include <moveit/planning_scene/planning_scene.h>
+#include <gtest/gtest.h>
+#include <thread>
+
+#include <moveit/collision_detection/collision_common.h>
+#include <moveit/collision_detection/collision_env.h>
+#include <moveit/collision_detection/collision_detector_allocator.h>
+
+const int TRIALS = 1000;
+const int THREADS = 2;
+
+bool runCollisionDetection(unsigned int id, unsigned int trials, const planning_scene::PlanningScene* scene,
+                           const robot_state::RobotState* state)
+{
+  collision_detection::CollisionRequest req;
+  collision_detection::CollisionResult res;
+  for (unsigned int i = 0; i < trials; ++i)
+  {
+    res.clear();
+    scene->checkCollision(req, res, *state);
+  }
+  return res.collision;
+}
+
+void runCollisionDetectionAssert(unsigned int id, unsigned int trials, const planning_scene::PlanningScene* scene,
+                                 const robot_state::RobotState* state, bool expected_result)
+{
+  ASSERT_EQ(expected_result, runCollisionDetection(id, trials, scene, state));
+}
+
+class CollisionDetectorThreadedTest : public testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    robot_model_ = moveit::core::loadTestingRobotModel("panda");
+    robot_model_ok_ = static_cast<bool>(robot_model_);
+    robot_state_.reset(new robot_state::RobotState(robot_model_));
+    planning_scene_.reset(new planning_scene::PlanningScene(robot_model_));
+  }
+
+  void TearDown() override
+  {
+  }
+
+  bool robot_model_ok_;
+
+  robot_model::RobotModelPtr robot_model_;
+
+  collision_detection::CollisionEnvPtr cenv_;
+
+  collision_detection::AllowedCollisionMatrixPtr acm_;
+
+  robot_state::RobotStatePtr robot_state_;
+
+  planning_scene::PlanningScenePtr planning_scene_;
+};
+
+TEST_F(CollisionDetectorThreadedTest, InitOk)
```

Comment:
To ensure that the robot_model ist valid, it doesn't make sense to define a separate test like this, as each test runs from its own test fixture instance. From the [official docs](https://github.com/google/googletest/blob/master/googletest/docs/primer.md):

> For each test defined with TEST_F(), googletest will create a fresh test fixture at runtime, immediately initialize it via SetUp(), run the test, clean up by calling TearDown(), and then delete the test fixture. Note that different tests in the same test suite have different test fixture objects, and googletest always deletes a test fixture before it creates the next one. googletest does not reuse the same test fixture for multiple tests. Any changes one test makes to the fixture do not affect other tests.

Reply:
But to ensure that in general this works? I assumed this to be good practice as this is done in all collision test files which existed like [here](https://github.com/ros-planning/moveit/blob/master/moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp)

Reply:
It doesn't make sense there as well. Better it would be to `ASSERT(robot_model_)` in `Setup()`.

Reply:
I made a PR #1777 to fix this. For now, just for this test as the other's will be changed to the templated version when `feature-bullet` is merged.